Implement MOEA/D - DE

jonpsy opened the pull request:
This PR is a continuation of #210. Till now I've made these changes

Address issue #176 
* Naming changes
* Fixing algorithms 
* Wrapping codes into functions
* Argument passing fixed

The following remains to be done


**TODO**
  

- [x] Adding more sanity checks
- [x] ~Let user explicitly choose whether they want to preserve diversity or not~ 
- [x] ~Use a Policy-based approach to delegate Weight initialization and Decomposition.~
- [x] Discard ```arma::mat``` for ```InputType``` and ```OutputType```.
- [x] Add test cases for various MatType: Stress testing and accuracy measurement. Schaffer and Fonseca pass currently for 
double.
- [x] Use the CheckArbitraryFunctionType API once that PR is merged.
- [x] Assert ```RequireDenseFloatingType``` InputType.
- [x] ~After ZDT is merged, test against it: report results. (or maybe do it in separate PR?)~

 **FIXME**
- [x] Handle zero weight condition in Tchebycheff decomposition
- [x] Updating Solutions 

**DOCUMENTATION**

- [x] Write about MOEA/D-DE in optimizer.md
- [x] Update History.MD
- [x]  Finalize docs.

**Memory**
- [x] Find and remove redundant & expensive matrix copy.
- [x] Find and remove leaks

**Testing**
- [x] Add callback test

**REVIEWS**
 Requesting @zoq @say4n @rcurtin  for reviews.

jonpsy commented:
There are a few compile-time errors, I'll be fixing that over-the-course. I'd like reviews on algorithmic correctness and semantics.

jonpsy commented:
Just have to fix the tests, a little bit of tinkering aaaand we should be done! I'd like to add other decomposition methods (Bi, Weighted) as part of separated PR so as to avoid bloating :)

jonpsy commented:
**Good news**:
 => Everything compiles
**Bad news:**
=> All the test fail
=> ```preserveDiversity``` when set to false leads to SIG memory violation.

@zoq, the actual paper preserves diversity by default, but pagmo2 allows user to turn off that feature. By "preserving diversity", I mean to say the optimizer produces a diverse set of population.

It does by the following way (as per the paper):
a) Algorithm chooses to sample from either the neighbor or whole population, but favors the "whole population" thus increasing diversity.
b) The child solutions (however good they might be) cannot replace a threshold number of parents. For ex: if ```maxReplace``` = 10, no more than 10 children would replace original solution.

The above-mentioned feature can be switched off by disabling ```preserveDiversity```, but I was wondering, is it overkill? Because paper didn't allow any choice and simply favored preserving diversity.

jonpsy commented:
All tests pass! @zoq  could you review this please? I was wondering we could try and compare NSGA2 & MOEA/D-DE in terms of speed, correctness.

shikharjan commented:
hey man are you going to apply for gsoc?

jonpsy commented:
Hi @shikharjan , this would help you started: https://www.mlpack.org/gsoc.html. To learn more about how to contribute to mlpack, you can check this [site]( https://github.com/mlpack/mlpack/blob/master/CONTRIBUTING.md. )

I'd recommend reaching out in the [IRC/Gitter](https://gitter.im/mlpack/mlpack?at=5df1c77b55d9392300f712a1) instead of here to receive help. Please note that comments on PR should be mainly related code itself. Thanks!

jonpsy commented:
Pushing some more.  I'll be explaining my choice of default args over the next few comments.

zoq commented on doc/optimizers.md:
```diff
@@ -1563,6 +1563,60 @@ optimizer.Optimize(f, coordinates);
+*An optimizer for arbitrary multi-objective functions.*
+MOEA/D-DE (Multi Objective Evolutionary Algorithm based on Decomposition - Differential Evolution) is a multi
+objective optimization algorithm. It decomposes a multiobjective optimization problem into a number of scalar
+optimization subproblems and optimizes them simultaneously. Each subproblem randomly use information from neighboring
```
```suggestion
optimization subproblems and optimizes them simultaneously. Each subproblem randomly uses information from neighboring
```

zoq commented on doc/optimizers.md:
```diff
@@ -1563,6 +1563,60 @@ optimizer.Optimize(f, coordinates);
+
+#### Constructors
+* `MOEAD()`
+* `MOEAD(`_`populationSize, numGeneration, crossoverProb, mutationProb, mutationStrength, neighborSize, distributionIndex, lowerBound,      neighborProb, differentialWeight, maxReplace, upperBound`_`)`
```
```suggestion
* `MOEAD(`_`populationSize, numGeneration, crossoverProb, mutationProb, mutationStrength, neighborSize, distributionIndex, lowerBound, neighborProb, differentialWeight, maxReplace, upperBound`_`)`
```

Remove some extra spaces.

zoq commented on doc/optimizers.md:
```diff
@@ -1563,6 +1563,60 @@ optimizer.Optimize(f, coordinates);
+
+| **type** | **name** | **description** | **default** |
+|----------|----------|-----------------|-------------|
+| `size_t` | **`populationSize`** | The number of candidates in the population. This should be at least 4 in size and a multiple of 4. | `100` |
```
Looks like not all parameters are listed here e.g. `numGeneration`.

zoq commented on include/ensmallen_bits/moead/moead.hpp:
```diff
@@ -0,0 +1,322 @@
+/**
+ * @file moead.hpp
+ * @authors Utkarsh Rai, Nanubala Gnana Sai
+ *
```
```suggestion
 * @author Utkarsh Rai
 * @author Nanubala Gnana Sai
 *
```

That's the format we used in other files.

zoq reviewed (commented):
Thanks, I'll take a closer look into the code over the next days, do you mind to go over the code and adjust the indentation according to the style guidelines.

shikharjan commented:
> Hi @shikharjan , this would help you started: https://www.mlpack.org/gsoc.html. To learn more about how to contribute to mlpack, you can check this [site](https://github.com/mlpack/mlpack/blob/master/CONTRIBUTING.md.)
>
> I'd recommend reaching out in the [IRC/Gitter](https://gitter.im/mlpack/mlpack?at=5df1c77b55d9392300f712a1) instead of here to receive help. Please note that comments on PR should be mainly related code itself. Thanks!

Sorry man im a noob to this and i wanted to work on Multi-objective Optimization in gsoc so just wanted to know that we dont apply for the same project

jonpsy commented:
Default args:

- **populationSize**: 150. For 2 objectives he's chosen N=300, so I took N=150, assuming single objective.

![N default](https://user-images.githubusercontent.com/45007169/111059997-c98ccd00-84bf-11eb-96cc-15428a26f74c.png)

- **neighborSize** : 20
- **neighborProb**: 0.9
- **maxReplace**: 2

![Parameter-Set1](https://user-images.githubusercontent.com/45007169/111060032-ffca4c80-84bf-11eb-91fc-51d52b0d608d.png)



Some notes:  The paper **recommends**  maxReplace << neighborSize
![nr-delta](https://user-images.githubusercontent.com/45007169/111060079-6b141e80-84c0-11eb-8ca1-59d5868277ba.png)



Finally

**crossOverProb**: 1.0
**differentialWeight**: 0.5
**distributionIndex**: 20
**mutationRate**: 1/numVariables

![Parameter-Set3](https://user-images.githubusercontent.com/45007169/111060108-9d258080-84c0-11eb-96ef-948f23c386f3.png)

jonpsy commented:
@zoq  Gentle nudge :)

zoq commented on include/ensmallen_bits/moead/moead.hpp:
```diff
@@ -0,0 +1,302 @@
+   * @param neighborSize A matrix containing indices of the neighbors.
+   * @return std::tuple<size_t, size_t> The chosen pair of indices.
+   */
+  std::tuple<size_t, size_t>
```
The type should come before the method name unless it's a typename.

zoq commented on include/ensmallen_bits/moead/moead_impl.hpp:
```diff
@@ -0,0 +1,453 @@
+                                            MatType& iterate,
+                                            CallbackTypes&&... callbacks)
+{
+  // Sanity checks
```
```suggestion
  // Sanity checks.
```

Missing stop at the end.

zoq commented on include/ensmallen_bits/moead/moead_impl.hpp:
```diff
@@ -0,0 +1,453 @@
+  arma::uvec shuffle;
+  // The weight matrix. Each vector represents a decomposition subproblem.
+  arma::Mat<ElemType> weights(numObjectives, populationSize, arma::fill::randu);
+  weights += 1E-10; // Numerical stability
```
Let us put that in the constructor, see https://github.com/mlpack/ensmallen/blob/master/include/ensmallen_bits/adam/adam.hpp#L98 for an example.

zoq replied:
I guess, still a good idea to add `epsilon` into the constructor and use that parameter here.

zoq commented on include/ensmallen_bits/moead/moead_impl.hpp:
```diff
@@ -0,0 +1,453 @@
+    distances =
+        arma::sqrt(arma::sum(arma::square(weights.col(i) - weights.each_col())));
+    arma::uvec sortedIndices = arma::stable_sort_index(distances);
+    // Ignore distance from self
```
```suggestion
    // Ignore distance from self.
```

Missing stop.

zoq commented on include/ensmallen_bits/moead/moead_impl.hpp:
```diff
@@ -0,0 +1,453 @@
+
+  // 1.2 Random generation of the initial population.
+  std::vector<MatType> population(populationSize);
+  for (size_t i = 0; i < populationSize; ++i)
+  {
+	  population[i] =
+		    arma::randu<MatType>(iterate.n_rows, iterate.n_cols) - 0.5 + iterate;
+
+	  for (size_t geneIdx = 0; geneIdx < numVariables; ++geneIdx)
+	  {
+		  if (population[i][geneIdx] < lowerBound[geneIdx])
+			  population[i][geneIdx] = lowerBound[geneIdx];
+		  else if (population[i][geneIdx] > upperBound[geneIdx])
+			  population[i][geneIdx] = upperBound[geneIdx];
+	  }
+  }
```
The indentation is off.

zoq commented on include/ensmallen_bits/moead/moead_impl.hpp:
```diff
@@ -0,0 +1,453 @@
+      size_t r1, r2, r3;
+      r1 = i;
+      // Randomly choose to sample from the population or the neighbors.
+      bool sampleNeighbor = ( arma::randu() < neighborProb );
```
```suggestion
      bool sampleNeighbor = arma::randu() < neighborProb;
```

No need for the extra `()`.

zoq commented on include/ensmallen_bits/moead/moead_impl.hpp:
```diff
@@ -0,0 +1,453 @@
+      double delta = arma::randu();
+      for (size_t geneIdx = 0; geneIdx < numVariables; ++geneIdx)
+      {
+
```
```suggestion
```

Remove the extra empty line here.

zoq commented on include/ensmallen_bits/moead/moead_impl.hpp:
```diff
@@ -0,0 +1,453 @@
+        if (delta < crossoverProb)
+        {
+          candidate[geneIdx] = population[r1][geneIdx] +
+              differentialWeight * (population[r2][geneIdx] -
+                                      population[r3][geneIdx]);
```
The indentation is off.

zoq commented on include/ensmallen_bits/moead/moead_impl.hpp:
```diff
@@ -0,0 +1,453 @@
+          {
+            candidate[geneIdx] = lowerBound[geneIdx] +
+                arma::randu() * (population[r1][geneIdx] -
+                                  lowerBound[geneIdx]);
+          }
+          if (candidate[geneIdx] > upperBound[geneIdx])
+          {
+            candidate[geneIdx] = upperBound[geneIdx] -
+                arma::randu() * (upperBound[geneIdx] -
+                                  population[r1][geneIdx]);
```
The indentation is off.

zoq commented on include/ensmallen_bits/moead/moead_impl.hpp:
```diff
@@ -0,0 +1,453 @@
+          candidate[geneIdx] = population[r1][geneIdx];
+        }
+
+	    Mutate(candidate, 1.0 / static_cast<double>(numVariables), lowerBound, upperBound);
```
The indentation is off.

zoq commented on include/ensmallen_bits/moead/moead_impl.hpp:
```diff
@@ -0,0 +1,453 @@
+
+        size_t pick = sampleNeighbor ? neighborIndices(idx, i) : idx;
+
+		    double candidateDecomposition = DecomposeObjectives(
+			      weights.col(pick), idealPoint, candidateFitness);
+		    double parentDecomposition = DecomposeObjectives(
+			      weights.col(pick), idealPoint, populationFitness.col(pick));
```
The indentation is off.

zoq commented on include/ensmallen_bits/moead/moead_impl.hpp:
```diff
@@ -0,0 +1,453 @@
+    const arma::umat& neighborIndices,
+    bool sampleNeighbor)
+{
+	size_t k, l;
+
+  k = sampleNeighbor
+		  ? neighborIndices(
+		        arma::randi(arma::distr_param(0, neighborSize - 1u)), subProblemIdx)
+		  : arma::randi(arma::distr_param(0, populationSize - 1u));
+
+  l = sampleNeighbor
+		  ? neighborIndices(
+		        arma::randi(arma::distr_param(0, neighborSize - 1u)), subProblemIdx)
+		  : arma::randi(arma::distr_param(0, populationSize - 1u));
+
+  if (k == l)
+	{
+		if (k == populationSize - 1u)
+			--k;
+    else
+			++k;
+	}
+
```
The indentation is off.

zoq commented on tests/moead_test.cpp:
```diff
@@ -0,0 +1,249 @@
+  const double upperBound = 10;
+
+  MOEAD opt(
+          150, // population size
+          1000,  // max generations
+          1.0,  // crossover prob
+          20, //neighbor size
+          20, //distribution index
+          0.9, //neighbor prob
+          0.5, //differential weight
+          2, //maxreplace,
+          lowerBound, //lower bound
+          upperBound //upper bound
```
Should start with a capital letter and end with a stop.

zoq commented on tests/moead_test.cpp:
```diff
@@ -0,0 +1,249 @@
+ */
+TEST_CASE("MOEADSchafferN1Test", "[MOEADTest]")
+{
+  SchafferFunctionN1<arma::mat> SCH;
```
Can you write the same tests for `arma::fmat` as well?

jonpsy replied:
This unearthed sooooo many problems. Thank you!

zoq reviewed (commented):
I have to take another look at the paper, but this looks really good.

jonpsy commented:
I was wondering why the indentation messes up when I push it here. Looks like tabs vs space issue, thanks for pointing them out @zoq.

jonpsy commented:
Stalling this for a while, till NSGA2 and Recursive MOO gets merged. They will serve as a baseline for this pull request. Have updated TODO list for others to track the progress of this PR.

jonpsy commented:
leave open

jonpsy commented:
Woops, sorry

zoq commented:
> Woops, sorry

You can close/reopen the PR as often as you like.

jonpsy commented:
Actually, I was checking this PR via my mobile and the "Close" button was quite shiny so I clicked it, haha.

jonpsy commented:
Okay, how about we have this merged and handle templating in another PR @zoq ?

zoq commented on include/ensmallen_bits/moead/moead.hpp:
```diff
@@ -0,0 +1,316 @@
+   * @param populationSize The number of elements in the population.
+   * @param maxGenerations The maximum number of generations allowed.
+   * @param crossoverProb The probability that a crossover will occur.
+   * @param neighborSize The number of nearest neighbours of weights
```
Maybe it makes sense to have `neighborProb` first followed by `neighborSize` instead of putting `distributionIndex` in between.

zoq commented on include/ensmallen_bits/moead/moead_impl.hpp:
```diff
@@ -0,0 +1,421 @@
+  if (lowerBound.n_rows == 1)
+    lowerBound = lowerBound(0, 0) * arma::ones(iterate.n_rows, iterate.n_cols);
+
+  // Check if lower bound is a vector of a single dimension.
```
```suggestion
  // Check if upper bound is a vector of a single dimension.
```

zoq commented on include/ensmallen_bits/moead/moead_impl.hpp:
```diff
@@ -0,0 +1,421 @@
+  for (BaseMatType& individual : population)
+  {
+      individual = arma::randu<BaseMatType>(
+          iterate.n_rows, iterate.n_cols) -0.5 + iterate;
```
```suggestion
          iterate.n_rows, iterate.n_cols) - 0.5 + iterate;
```

Pedantic style issue.

zoq commented on include/ensmallen_bits/moead/moead_impl.hpp:
```diff
@@ -0,0 +1,421 @@
+                arma::randu() * (castedUpperBound(geneIdx) - population[r1](geneIdx));
+          }
+        }
+
```
```suggestion
```

Remove the extra empty line here.

zoq commented on include/ensmallen_bits/moead/moead_impl.hpp:
```diff
@@ -0,0 +1,421 @@
+      double delta = arma::randu();
+      for (size_t geneIdx = 0; geneIdx < numVariables; ++geneIdx)
+      {
+        if (delta < crossoverProb)
```
Maybe I missed something but shouldn't we use another crossover probability  for each gene?

jonpsy replied:
I think you mean delta should be generated inside the loop? That's a correct remark, I'll change that.

zoq commented on include/ensmallen_bits/moead/moead_impl.hpp:
```diff
@@ -0,0 +1,421 @@
+        if (replaceCounter >= maxReplace)
+          break;
+
+        size_t pick = sampleNeighbor ? neighborIndices(idx, subProblemIdx) : idx;
```
Could be `const` to make it obvious that the value will not change.

zoq commented on include/ensmallen_bits/moead/moead_impl.hpp:
```diff
@@ -0,0 +1,421 @@
+            arma::randi(arma::distr_param(0, neighborSize - 1u)), subProblemIdx)
+      : arma::randi(arma::distr_param(0, populationSize - 1u));
+
+  if (k == l)
```
Would be nice to add a comment here.

zoq commented on include/ensmallen_bits/moead/moead_impl.hpp:
```diff
@@ -0,0 +1,421 @@
+                          const MatType& upperBound)
+{
+    size_t numVariables = candidate.n_rows;
+    for (size_t geneIdx=0; geneIdx < numVariables; ++geneIdx)
```
```suggestion
    for (size_t geneIdx = 0; geneIdx < numVariables; ++geneIdx)
```

zoq commented on include/ensmallen_bits/moead/moead_impl.hpp:
```diff
@@ -0,0 +1,421 @@
+      double value, perturbationFactor;
+      if(rand < 0.5)
+      {
+        value = 2. * rand + (1. - 2. * rand) *
+            std::pow(upperDelta, distributionIndex + 1.0);
+        perturbationFactor = std::pow(value, mutationPower) - 1.;
```
```suggestion
        value = 2.0 * rand + (1.0 - 2. * rand) *
            std::pow(upperDelta, distributionIndex + 1.0);
        perturbationFactor = std::pow(value, mutationPower) - 1.0;
```

zoq commented on include/ensmallen_bits/moead/moead_impl.hpp:
```diff
@@ -0,0 +1,421 @@
+      }
+      else
+      {
+        value = 2. * (1. - rand) + 2.*(rand - 0.5) *
+            std::pow(lowerDelta, distributionIndex + 1.0);
+        perturbationFactor = 1. - std::pow(value, mutationPower);
```
```suggestion
        value = 2.0 * (1.0 - rand) + 2.0 *(rand - 0.5) *
            std::pow(lowerDelta, distributionIndex + 1.0);
        perturbationFactor = 1.0 - std::pow(value, mutationPower);
```

jonpsy commented:
@zoq  I think I've addressed all of your comments. If you find some indent mistakes, it would be helpful if you could do a  "commit suggestion", it gets tiring to change minor things every commits. Thanks a lot.

jonpsy commented:
Uhm... I've done a sizeable amount of work on the policy-based method already. How about we merge this as-is now, and create a follow-up PR in which I handle the templatizing and some other minor fixes.

zoq commented:
> @zoq I think I've addressed all of your comments. If you find some indent mistakes, it would be helpful if you could do a "commit suggestion", it gets tiring to change minor things every commits. Thanks a lot.

Maybe you can adjust your code editor to use the correct indentation style, I use vim and setting up the indentation is easy, That could save us both some time. https://github.com/mlpack/mlpack/wiki/DesignGuidelines has all the important style guidelines.

zoq commented:
> Uhm... I've done a sizeable amount of work on the policy-based method already. How about we merge this as-is now, and create a follow-up PR in which I handle the templatizing and some other minor fixes.
> 
> Here's a [patch](https://drive.google.com/file/d/1TjLNWouEA8IZVjVBti7XBUDHOdsFlEab/view?usp=sharing) for policy if you want a look.

You can just open another PR with the changes from this one and the policy design changes, and rebase the PR afterwards. I don't want to rush anything. Like one thing I haven't done is run the tests in a loop and check if they are stable, maybe you did that already in which case we can check that one.

zoq commented on include/ensmallen_bits/moead/moead_impl.hpp:
```diff
@@ -0,0 +1,430 @@
+      idealPoint = arma::min(idealPoint, candidateFitness);
+
+      // 2.5 Update of the population.
+      size_t replaceCounter {0};
```
```suggestion
      size_t replaceCounter = 0;
```

To be consistent.

zoq commented on include/ensmallen_bits/moead/moead_impl.hpp:
```diff
@@ -0,0 +1,430 @@
+  for (BaseMatType& individual : population)
+  {
+  	individual = arma::randu<BaseMatType>(
+    		iterate.n_rows, iterate.n_cols) - 0.5 + iterate;
```
```suggestion
        iterate.n_rows, iterate.n_cols) - 0.5 + iterate;
```

zoq commented on include/ensmallen_bits/moead/moead.hpp:
```diff
@@ -0,0 +1,326 @@
+   * @param neighborSize A matrix containing indices of the neighbors.
+   * @return std::tuple<size_t, size_t> The chosen pair of indices.
+   */
+  std::tuple<size_t, size_t> Mating(size_t subProblemIdx,
+                                             const arma::umat& neighborSize,
+                                             bool sampleNeighbor);
```
```suggestion
  std::tuple<size_t, size_t> Mating(size_t subProblemIdx,
                                     const arma::umat& neighborSize,
                                     bool sampleNeighbor);
```

zoq reviewed (commented):
Next step to test this in a loop.

jonpsy commented:
Thank you for taking the time to fix the style. I'll be sure to change my IDE settings.

Reg testing, one thing you'll find interesting to know is, that higher population size leads to more accurate results.

 I have yet to test this claim, I'll play around some parameters run it over a 1000 times and report the results here.

zoq commented:
> Reg testing, one thing you'll find interesting to know is, that higher population size leads to more accurate results.

That is true, we basically increase the number of possible solutions. For the tests we don't need a perfect solution, just something that is close, to check that the optimizer works with different starting solutions.

jonpsy commented:
Ran these command ```ctest -C Release -V --output-on-failure --repeat-until-fail 1000 --rng-seed time``` to test over 1000 times. It passed each and every time. [link](https://pastebin.com/8cC4MinQ). Feel free (and I encourage you to) try this yourself. Just comment out other tests from ```CMakeLists.txt```.  

Let's do some final cleanups. Doc fixes, if any, and then this should be good to go imo. Looking forward for your review.

zoq reviewed (approved):
Great work, no further comments from my side.

jonpsy commented:
All right, if nobody has any objections. Let's get this merged!

zoq commented:
Nice, with that one merged, let's open another PR for the policy design changes.
